File: backend/src/StajProject.Domain/Common/PermissionCodes.cs
namespace StajProject.Domain.Common;

/// <summary>
/// Kanonik yetki kodları. Yetkilendirme <b>yalnızca</b> bu kodlar üzerinden
/// çalışır; koda dağılmış <c>"drawings.delete"</c> string'leri kullanılmaz ve
/// Türkçe görünen adlar ASLA kimlik olarak kullanılmaz.
/// </summary>
/// <remarks>
/// <para>
/// <b>Kodlar değiştirilemez.</b> Veritabanındaki <c>permissions.code</c>
/// satırları ve onlara bağlı grant kayıtları bu değerlere göre eşleşir; bir
/// kodun yeniden adlandırılması, o yetkiye sahip herkesin yetkisini sessizce
/// kaybetmesi demektir.
/// </para>
/// <para>
/// <b>Kapsam koda gömülmez.</b> <c>drawings.view.own</c> / <c>drawings.view.all</c>
/// gibi türevler bilinçli olarak YOKTUR — kayıt kapsamı ayrı bir eksendir ve
/// sonraki bir fazda ele alınır.
/// </para>
/// </remarks>
public static class PermissionCodes
{
    /* --- Harita ---------------------------------------------------------------- */

    public const string MapView = "map.view";

    /* --- Çizim oluşturma ------------------------------------------------------- */

    public const string DrawingsPointCreate = "drawings.point.create";
    public const string DrawingsLineCreate = "drawings.line.create";
    public const string DrawingsPolygonCreate = "drawings.polygon.create";

    /* --- Çizim yönetimi -------------------------------------------------------- */

    public const string DrawingsView = "drawings.view";
    public const string DrawingsMetadataUpdate = "drawings.metadata.update";
    public const string DrawingsGeometryUpdate = "drawings.geometry.update";
    public const string DrawingsStyleUpdate = "drawings.style.update";
    public const string DrawingsDelete = "drawings.delete";
    public const string DrawingsRestore = "drawings.restore";

    /* --- Araçlar --------------------------------------------------------------- */

    public const string MeasurementUse = "measurement.use";
    public const string SelectionUse = "selection.use";

    /* --- Envanter -------------------------------------------------------------- */

    public const string InventoryView = "inventory.view";
    public const string InventoryAnalysis = "inventory.analysis";

    /* --- Isı haritası ----------------------------------------------------------- */

    /* Isı haritası kendi yetkisidir ve inventory.analysis altına GİZLENMEZ:
       envanter analizi kesişim/dağılım sorgularını çalıştırmaktır; ısı haritası
       ise kişinin kendi kayıtlarının yoğunluğunu, sunucuda üretilmiş bir WMS
       görüntüsü olarak görmektir. İki yetenek ayrı verilebilmelidir — birine
       sahip olmak diğerini ima etmez. */

    public const string HeatmapView = "heatmap.view";

    /* --- Konum analizi ---------------------------------------------------------- */

    /* Konum analizi ÜÇÜNCÜ bir yetenektir ve ne inventory.analysis ne de
       heatmap.view altına gizlenir: envanter analizi kişinin KENDİ çizim
       envanteriyle bir alanın kesişimini sayar; ısı haritası yine KENDİ
       noktalarının yoğunluğunu gösterir; konum analizi ise ORTAK POI
       envanterini, kullanıcının seçtiği kategori ağırlıklarıyla puanlar.
       Üçünün verisi de, kapsamı da farklıdır — birine sahip olmak diğerini
       ima etmemelidir.

       Kod alan-önce/yetenek-sonra yazılır (inventory.analysis ile aynı
       biçim): ilk parça alanı, son parça yeteneği söyler. */

    public const string LocationAnalysis = "location.analysis";

    /* --- Katmanlar ------------------------------------------------------------- */

    public const string LayersView = "layers.view";
    public const string LayersManage = "layers.manage";

    /* --- Kullanıcılar ---------------------------------------------------------- */

    public const string UsersView = "users.view";
    public const string UsersCreate = "users.create";
    public const string UsersUpdate = "users.update";
    public const string UsersDeactivate = "users.deactivate";
    public const string UsersDelete = "users.delete";

    /* --- Roller ---------------------------------------------------------------- */

    public const string RolesView = "roles.view";
    public const string RolesCreate = "roles.create";
    public const string RolesUpdate = "roles.update";
    public const string RolesDelete = "roles.delete";

    /* --- Yetkiler -------------------------------------------------------------- */

    public const string PermissionsView = "permissions.view";
    public const string PermissionsAssign = "permissions.assign";

    /* --- Coğrafi yetkilendirme ------------------------------------------------- */

    /* Coğrafi yetki alanı yönetimi ayrı bir güvenlik yeteneğidir ve bilinçli
       olarak users.update / roles.update altına GİZLENMEZ: coğrafi alan
       yalnızca bir kullanıcı alanı değil, o kullanıcının nerede veri
       üretebileceğini belirleyen bir sınırdır. Kullanıcı düzenleme yetkisinin
       sessizce bu sınırı da kaldırabilmesi, yetki yükseltmeye açık kapı
       bırakırdı. */

    public const string GeographyView = "geography.view";
    public const string GeographyManage = "geography.manage";

    /* --- Denetim --------------------------------------------------------------- */

    /* Aktivite geçmişi kendi yetkisidir ve users.view / permissions.view altına
       GİZLENMEZ: kayıt, kimin neyi ne zaman değiştirdiğini gösterir — yani
       diğer yöneticilerin hareketlerini de. Kullanıcı listesini görebilmek,
       yönetim geçmişini okuyabilmekle aynı şey değildir. */

    public const string ActivityView = "activity.view";

    /* --- POI --------------------------------------------------------------------

       POI yetkileri çizim yetkilerinden AYRIDIR ve onların altına gizlenmez:
       POI, stil taşımayan, sahibine göre gizlenmeyen ve kategori hiyerarşisine
       bağlı ORTAK bir envanterdir. Bir kişinin kendi çizimlerini yönetebilmesi,
       herkesin gördüğü POI envanterine kayıt ekleyebilmesiyle aynı şey
       değildir.

       Görüntüleme ve yönetim de ayrılır: haritada pin görmek ile kimin neyi
       eklediğini listeleyebilmek farklı yeteneklerdir (activity.view ile aynı
       gerekçe). Kategori yönetimi ise üçüncü bir eksendir — herkesin
       sınıflandırma yapmak zorunda olduğu taksonomiyi tanımlamak, kayıtları
       gözden geçirmekten başka bir otoritedir. */

    public const string PoiView = "poi.view";
    public const string PoiCreate = "poi.create";

    /* Güncelleme ve silme, oluşturmadan AYRI kodlardır ve `poi.create` üzerine
       yüklenmez: bir noktayı eklemek ile başkasının da göreceği ortak
       envanterdeki bir kaydı değiştirmek/kaldırmak farklı yeteneklerdir.

       İkisi de SAHİPLİK ile birlikte okunur — kod "kendi kaydında" yetki
       verir; herkesin kaydına yetki veren tek şey `poi.manage`'dir. Bu ayrım
       kodun kendisinde değil, servis katmanındaki sahiplik denetimindedir:
       yetki "ne yapabilir", sahiplik "hangi kayıtta" sorusunu yanıtlar. */

    public const string PoiUpdate = "poi.update";
    public const string PoiDelete = "poi.delete";

    public const string PoiManage = "poi.manage";
    public const string PoiCategoriesManage = "poi.categories.manage";

    /* --- Akıllı ulaşım -------------------------------------------------------- */

    public const string TransportView = "transport.view";

    public const string TransportStopCreate = "transport.stop.create";
    public const string TransportStopUpdate = "transport.stop.update";
    public const string TransportStopDelete = "transport.stop.delete";
    public const string TransportStopRestore = "transport.stop.restore";

    public const string TransportRouteCreate = "transport.route.create";
    public const string TransportRouteUpdate = "transport.route.update";
    public const string TransportRouteDelete = "transport.route.delete";
    public const string TransportRouteRestore = "transport.route.restore";
    public const string TransportRouteReorder = "transport.route.reorder";

    /* Simülasyon başlatmak, rota verisini DEĞİŞTİRMEZ: bu yüzden
       `transport.route.update` üzerine yüklenmez, ayrı bir kod olur. Ayrım
       operasyoneldir — bir kullanıcı hattı canlı işletebilirken güzergahın
       kendisini yeniden hesaplayamayabilir; tersi de mümkündür.

       Görüntüleme yetkisi de İMA ETMEZ: haritada aracın nerede olduğunu
       görmek `transport.view` ile olur, hattı çalıştırmak bu kodla. */
    public const string TransportSimulationStart = "transport.simulation.start";

    /// <summary>EF <c>HasMaxLength</c> ile aynı sınır.</summary>
    public const int MaxLength = 128;
}

File: frontend/src/auth/permissionCodes.js
/**
 * Kanonik yetki kodları — backend'in `PermissionCodes` sabitlerinin birebir
 * karşılığı.
 *
 * <b>Kaynak burası DEĞİLDİR.</b> Gerçek katalog veritabanındadır ve
 * yetkilendirme kararını her zaman backend verir. Bu dosya yalnızca kodların
 * arayüzde harf harf yazılmasını önler: `can('drawings.point.creat')` gibi bir
 * yazım hatası sessizce "yetki yok" demek olurdu ve hiçbir yerde hata
 * vermezdi.
 *
 * <b>Burada rol yoktur.</b> "Viewer şunları görür" biçiminde bir matris
 * bilinçli olarak eklenmez: rol→yetki eşlemesinin sahibi veritabanıdır ve onu
 * burada ikinci kez yazmak, bir rolün yetkisi değiştiği gün sessizce çelişen
 * iki kural kitabı yaratırdı. Arayüz yalnızca ETKİN kodları tüketir; özel bir
 * rol ya da kullanıcıya özel bir yetki bu sayede kendiliğinden çalışır.
 */
export const PERMISSIONS = Object.freeze({
  /* --- Harita ------------------------------------------------------------- */
  MAP_VIEW: 'map.view',

  /* --- Çizim oluşturma ---------------------------------------------------- */
  DRAWINGS_POINT_CREATE: 'drawings.point.create',
  DRAWINGS_LINE_CREATE: 'drawings.line.create',
  DRAWINGS_POLYGON_CREATE: 'drawings.polygon.create',

  /* --- Çizim yönetimi ----------------------------------------------------- */
  DRAWINGS_VIEW: 'drawings.view',
  DRAWINGS_METADATA_UPDATE: 'drawings.metadata.update',
  DRAWINGS_GEOMETRY_UPDATE: 'drawings.geometry.update',
  DRAWINGS_STYLE_UPDATE: 'drawings.style.update',
  DRAWINGS_DELETE: 'drawings.delete',
  DRAWINGS_RESTORE: 'drawings.restore',

  /* --- Araçlar ------------------------------------------------------------ */
  MEASUREMENT_USE: 'measurement.use',
  SELECTION_USE: 'selection.use',

  /* --- Envanter ----------------------------------------------------------- */
  INVENTORY_VIEW: 'inventory.view',
  INVENTORY_ANALYSIS: 'inventory.analysis',

  /* --- Isı haritası -------------------------------------------------------- */
  /* Envanter analizinden AYRI bir koddur: analiz çalıştırabilen birinin ısı
     haritasını da görebildiği varsayımı bilinçli olarak yoktur. */
  HEATMAP_VIEW: 'heatmap.view',

  /* --- Katmanlar ---------------------------------------------------------- */
  LAYERS_VIEW: 'layers.view',
  LAYERS_MANAGE: 'layers.manage',

  /* --- Kullanıcılar ------------------------------------------------------- */
  USERS_VIEW: 'users.view',
  USERS_CREATE: 'users.create',
  USERS_UPDATE: 'users.update',
  USERS_DEACTIVATE: 'users.deactivate',
  USERS_DELETE: 'users.delete',

  /* --- Roller ------------------------------------------------------------- */
  ROLES_VIEW: 'roles.view',
  ROLES_CREATE: 'roles.create',
  ROLES_UPDATE: 'roles.update',
  ROLES_DELETE: 'roles.delete',

  /* --- Yetkiler ----------------------------------------------------------- */
  PERMISSIONS_VIEW: 'permissions.view',
  PERMISSIONS_ASSIGN: 'permissions.assign',

  /* --- Coğrafi yetkilendirme ---------------------------------------------- */
  /* Katalog eksiksiz tutulur: bu kodları HENÜZ hiçbir ekran tüketmiyor —
     coğrafi yetki arayüzü sonraki fazda gelir. Sabitin burada durması, o ekran
     yazılırken kodun elle yeniden yazılmasını (ve sessizce yanlış yazılmasını)
     önler. */
  GEOGRAPHY_VIEW: 'geography.view',
  GEOGRAPHY_MANAGE: 'geography.manage',

  /* --- Denetim ------------------------------------------------------------ */
  /* Aktivite geçmişi kendi yetkisidir ve users.view / permissions.view altına
     gizlenmez: kayıt, diğer yöneticilerin hareketlerini de gösterir. Kullanıcı
     listesini görebilmek, yönetim geçmişini okuyabilmekle aynı şey değildir. */
  ACTIVITY_VIEW: 'activity.view',
})

/**
 * Çizim türü → o türü OLUŞTURMA yetkisi.
 *
 * Üç kod ayrı kalır ve tek bir "çizim oluşturma" yetkisine indirgenmez, çünkü
 * backend'de de ayrılar: `POST /api/drawings/point` yalnızca
 * `drawings.point.create` arar. Geçmiş yığınındaki bir "ileri al" adımı da
 * aynı ucu çağırdığı için aynı eşlemeyi kullanır.
 */
export const DRAWING_CREATE_PERMISSIONS = Object.freeze({
  point: PERMISSIONS.DRAWINGS_POINT_CREATE,
  line: PERMISSIONS.DRAWINGS_LINE_CREATE,
  polygon: PERMISSIONS.DRAWINGS_POLYGON_CREATE,
})

/**
 * Yönetim panelinin bölümleri, kenar çubuğundaki SIRAYLA.
 *
 * Tek tanım: kenar çubuğu, `/admin` kök yönlendirmesi ve rota koruyucuları
 * hepsi bunu okur. Üç yerde üç ayrı liste tutmak, bir bölüm eklendiğinde
 * menüde görünüp yönlendirmede atlanmasına açık kapı bırakırdı.
 */
export const ADMIN_SECTIONS = Object.freeze([
  { path: '/admin/users', permission: PERMISSIONS.USERS_VIEW },
  { path: '/admin/roles', permission: PERMISSIONS.ROLES_VIEW },
  { path: '/admin/permissions', permission: PERMISSIONS.PERMISSIONS_VIEW },
  /* Aktivite geçmişi de bir yönetim BÖLÜMÜDÜR: yalnızca `activity.view`
     taşıyan bir denetçi, başka hiçbir yetkisi olmasa bile yönetim panelini
     açabilmeli ve doğrudan bu sayfaya yönlendirilmelidir. Listeye eklenmeseydi
     menüde görünür ama /admin kökü onu hiç seçmezdi. */
  { path: '/admin/activity', permission: PERMISSIONS.ACTIVITY_VIEW },
])

/** Yönetim paneline girişi açan yetkiler: en az biri yeterlidir. */
export const ADMIN_ENTRY_PERMISSIONS = Object.freeze(ADMIN_SECTIONS.map((section) => section.permission))
